File: native-plugins/admob/resources/ios/Sources/AppOpenLifecycle.swift
import Foundation
import UIKit
import GoogleMobileAds

/// Auto-show app-open ads when the app foregrounds via
/// `UIApplication.didBecomeActiveNotification`, after skipping the first
/// resume (cold-start splash). Honours the 4-hour staleness rule via
/// `AppOpenRegistry.shared.isFresh()`.
///
/// Stale ads are silently discarded; a fresh load() is NOT kicked off here -
/// that's the consumer's job.
///
/// Dormant in Buff: app-open bridges and registration are intentionally absent
/// from the plugin manifest.
enum AppOpenLifecycle {
    private static var registered = false
    // iOS init runs after the app finishes launching; cold-start
    // didBecomeActive has already fired by then. Same reasoning as Android -
    // default to "consumed" so the first observed foreground triggers auto-show.
    private static var coldStartConsumed = true
    // Host-controlled kill-switch for the auto-show (set true while a user holds
    // an ad-free pass). Resets on process restart; the host re-syncs at boot.
    static var autoShowSuppressed = false

    static func register() {
        guard !registered else { return }
        registered = true

        NotificationCenter.default.addObserver(
            forName: UIApplication.didBecomeActiveNotification,
            object: nil,
            queue: .main
        ) { _ in
            if !coldStartConsumed {
                coldStartConsumed = true
                return
            }
            if autoShowSuppressed { return }
            if FullScreenAdState.recentlyDismissed() {
                // Another full-screen ad just dismissed - suppress to avoid back-to-back presentations.
                return
            }
            for slot in AppOpenRegistry.shared.allSlots() {
                guard let ad = AppOpenRegistry.shared.get(slot: slot) else { continue }
                if !AppOpenRegistry.shared.isFresh(slot: slot) {
                    AppOpenRegistry.shared.remove(slot: slot)
                    continue
                }
                guard let root = AdmobFunctions.rootViewController() else { continue }
                ad.present(from: root)
            }
        }
    }
}

File: native-plugins/admob/resources/android/src/AppOpenLifecycle.kt
package com.blessedzulu.nativephp.admob

import android.util.Log
import androidx.fragment.app.FragmentActivity
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.FullScreenContentCallback
import com.nativephp.mobile.lifecycle.NativePHPLifecycle
import com.nativephp.mobile.utils.NativeActionCoordinator
import org.json.JSONObject
import java.lang.ref.WeakReference

/**
 * Auto-show app-open ads when the app foregrounds, after skipping the first
 * resume (which is the cold-start splash). Honours the 4-hour staleness rule
 * via AppOpenRegistry.isFresh().
 *
 * Stale ads are silently discarded; a fresh load() is NOT kicked off here -
 * that's the consumer's job. Documented as a known limitation.
 *
 * Dormant in Buff: app-open bridges and registration are intentionally absent
 * from the plugin manifest.
 */
object AppOpenLifecycle {
    private var registered = false
    // Default to "consumed" so a host that deliberately registers this dormant
    // lifecycle would treat its first observed resume as a warm foreground.
    private var coldStartConsumed = true
    private var activityRef: WeakReference<FragmentActivity>? = null

    private const val EVENT_BASE = "BlessedZulu\\NativePhpAdmob\\Events"

    // Host-controlled kill-switch for the auto-show. Set true to stand down (e.g.
    // while the user holds a temporary ad-free pass) - the auto-show fires outside
    // any per-request gate, so it must be told to. Defaults false and resets on
    // process restart, so the host re-syncs it at boot.
    @JvmStatic
    var autoShowSuppressed = false

    @JvmStatic
    fun bindActivity(activity: FragmentActivity) {
        activityRef = WeakReference(activity)
    }

    @JvmStatic
    @Synchronized
    fun register() {
        if (registered) return
        registered = true

        NativePHPLifecycle.on("onResume") {
            Log.i("AppOpenLifecycle", "onResume callback fired (coldStartConsumed=$coldStartConsumed, slots=${AppOpenRegistry.allSlots().size}, activity=${activityRef?.get() != null}, recentlyDismissed=${FullScreenAdState.recentlyDismissed()})")
            if (!coldStartConsumed) {
                coldStartConsumed = true
                return@on
            }
            if (autoShowSuppressed) {
                Log.i("AppOpenLifecycle", "suppressing auto-show: disabled by host (e.g. ad-free pass)")
                return@on
            }
            if (FullScreenAdState.recentlyDismissed()) {
                Log.i("AppOpenLifecycle", "suppressing auto-show: another full-screen ad dismissed within ${FullScreenAdState.DISMISS_GRACE_MS}ms")
                return@on
            }
            val activity = activityRef?.get() ?: run {
                Log.w("AppOpenLifecycle", "activityRef null, skipping auto-show")
                return@on
            }
            activity.runOnUiThread {
                AppOpenRegistry.allSlots().forEach { slot ->
                    val ad = AppOpenRegistry.get(slot) ?: return@forEach
                    if (!AppOpenRegistry.isFresh(slot)) {
                        AppOpenRegistry.remove(slot)
                        return@forEach
                    }
                    Log.i("AppOpenLifecycle", "auto-showing slot=$slot")
                    attachCallback(activity, slot, ad)
                    ad.show(activity)
                }
            }
        }
        Log.i("AppOpenLifecycle", "registered onResume observer")
    }

    /**
     * Shared FullScreenContentCallback wiring used by both ShowAppOpen
     * (manual override) and the auto-show lifecycle path. Dispatches the
     * five standard lifecycle events with format=app_open. On dismissal or
     * failed-show, clears the registry slot.
     */
    @JvmStatic
    fun attachCallback(activity: FragmentActivity, slot: String, ad: com.google.android.gms.ads.appopen.AppOpenAd) {
        ad.fullScreenContentCallback = object : FullScreenContentCallback() {
            override fun onAdShowedFullScreenContent() {
                dispatch(activity, "AdShown", mapOf("slot" to slot, "format" to "app_open"))
            }

            override fun onAdDismissedFullScreenContent() {
                FullScreenAdState.markDismissed()
                AppOpenRegistry.remove(slot)
                dispatch(activity, "AdDismissed", mapOf("slot" to slot, "format" to "app_open"))
            }

            override fun onAdFailedToShowFullScreenContent(error: AdError) {
                FullScreenAdState.markDismissed()
                AppOpenRegistry.remove(slot)
                dispatch(
                    activity,
                    "AdFailedToShow",
                    mapOf(
                        "slot" to slot,
                        "format" to "app_open",
                        "errorCode" to error.code,
                        "errorMessage" to error.message,
                    ),
                )
            }

            override fun onAdImpression() {
                dispatch(activity, "AdImpression", mapOf("slot" to slot, "format" to "app_open"))
            }

            override fun onAdClicked() {
                dispatch(activity, "AdClicked", mapOf("slot" to slot, "format" to "app_open"))
            }
        }
    }

    private fun dispatch(activity: FragmentActivity, eventClass: String, payload: Map<String, Any>) {
        activity.runOnUiThread {
            try {
                NativeActionCoordinator.dispatchEvent(
                    activity,
                    "$EVENT_BASE\\$eventClass",
                    JSONObject(payload).toString(),
                )
            } catch (_: Throwable) {
                // swallowed; lifecycle observer must never crash the app
            }
        }
    }
}
